hbase-common/src/main/java/org/apache/hadoop/hbase/io/crypto/tls/X509Util.java

```diff
@@ -395,4 +411,57 @@ private static String[] getCipherSuites(Configuration config) {
       return cipherSuitesInput.split(",");
     }
   }
+
+  private static FileChangeWatcher newFileChangeWatcher(String fileLocation, Runnable resetContext)
+    throws IOException {
+    if (fileLocation == null || fileLocation.isEmpty() || resetContext == null) {
+      return null;
+    }
+    final Path filePath = Paths.get(fileLocation).toAbsolutePath();
+    Path parentPath = filePath.getParent();
+    if (parentPath == null) {
+      throw new IOException("Key/trust store path does not have a parent: " + filePath);
+    }
+    FileChangeWatcher fileChangeWatcher = new FileChangeWatcher(parentPath, watchEvent -> {
+      handleWatchEvent(filePath, watchEvent, resetContext);
+    });
+    fileChangeWatcher.start();
+    return fileChangeWatcher;
+  }
+
+  /**
+   * Handler for watch events that let us know a file we may care about has changed on disk.
+   * @param filePath the path to the file we are watching for changes.
+   * @param event    the WatchEvent.
+   */
+  private static void handleWatchEvent(Path filePath, WatchEvent<?> event, Runnable resetContext) {
+    boolean shouldResetContext = false;
+    Path dirPath = filePath.getParent();
+    if (event.kind().equals(StandardWatchEventKinds.OVERFLOW)) {
+      // If we get notified about possibly missed events, reload the key store / trust store just to
+      // be sure.
+      shouldResetContext = true;
+    } else if (
+      event.kind().equals(StandardWatchEventKinds.ENTRY_MODIFY)
+        || event.kind().equals(StandardWatchEventKinds.ENTRY_CREATE)
+    ) {
+      Path eventFilePath = dirPath.resolve((Path) event.context());
+      if (filePath.equals(eventFilePath)) {
+        shouldResetContext = true;
+      }
+    }
+    // Note: we don't care about delete events
+    if (shouldResetContext) {
+      if (LOG.isDebugEnabled()) {
```

Comment:
nit: Maybe worth info logging? My thought here is that this wouldn't be a frequent event, yet a important one to get logged at higher level than debugging.